daemon/snapper/snapname/snapname.go

```diff
@@ -0,0 +1,40 @@
+package snapname
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/zrepl/zrepl/daemon/snapper/snapname/timestamp"
+	"github.com/zrepl/zrepl/zfs"
+)
+
+type Formatter struct {
+	prefix    string
+	timestamp *timestamp.Formatter
+}
+
+func New(prefix, tsFormat, tsLocation string) (*Formatter, error) {
+	timestamp, err := timestamp.New(tsFormat, tsLocation)
+	if err != nil {
+		return nil, err
+	}
+	formatter := &Formatter{
+		prefix:    prefix,
+		timestamp: timestamp,
+	}
+	// Best-effort check to detect whether the result would be an invalid name early.
+	testFormat := "some/dataset@" + formatter.Format(time.Now())
+	if err := zfs.EntityNamecheck(testFormat, zfs.EntityTypeSnapshot); err != nil {
+		return nil, errors.Wrapf(err, "prefix or timestamp format result in invalid snapshot name such as %q", testFormat)
+	}
```

Comment:
This check isn't 100% correct / selft-contained.

For example, imagine out `timestamp.Formatter` didn't have the special handling for `+` timezone offsets.

If you're in the UK, the `+` appears only in the summer (BST is UTC+1)

=> The check would only fail in the summer. But we'd want it to fail all the time.